cep-16.md

```diff
@@ -0,0 +1,260 @@
+
+### Repodata shard
+
+Individual shards are stored under the URL `<shard_base_url>/<subdir>/shards/<sha256>.msgpack.zst`. Where the `sha256` is the lower-case hex representation of the bytes from the index. It is a zstandard compressed msgpack file that contains the metadata of the package. The `<shard_base_url>` is defined in [Authentication](#authentication).
+
+The files are content-addressable which makes them ideal to be served through a CDN. They SHOULD be served with `Cache-Control: immutable` header.
+
+The shard contains the repodata information that would otherwise have been found in the `repodata.json` file. 
+It is a dictionary that contains the following keys:
+
+**Example (written in JSON for readability):**
+
+```js
+{
+  // dictionary of .tar.bz2 files
+  "packages": {
+    "rich-10.15.2-pyhd8ed1ab_1.tar.bz2": {
+      "build": "pyhd8ed1ab_1",
+      "build_number": 1,
+      "depends": [
+        "colorama >=0.4.0,<0.5.0",
+        "commonmark >=0.9.0,<0.10.0",
+        "dataclasses >=0.7,<0.9",
+        "pygments >=2.6.0,<3.0.0",
+        "python >=3.6.2",
+        "typing_extensions >=3.7.4,<5.0.0"
+      ],
+      "license": "MIT",
+      "license_family": "MIT",
+      "md5": "2456071b5d040cba000f72ced5c72032",
+      "name": "rich",
+      "noarch": "python",
+      "sha256": "a38347390191fd3e60b17204f2f6470a013ec8753e1c2e8c9a892683f59c3e40",
+      "size": 153963,
+      "subdir": "noarch",
+      "timestamp": 1638891318904,
+      "version": "10.15.2"
+    }
+  },
+ // dictionary of .conda files
+  "packages.conda": {
+    "rich-13.7.1-pyhd8ed1ab_0.conda": {
+      "build": "pyhd8ed1ab_0",
+      "build_number": 0,
+      "depends": [
+        "markdown-it-py >=2.2.0",
+        "pygments >=2.13.0,<3.0.0",
+        "python >=3.7.0",
+        "typing_extensions >=4.0.0,<5.0.0"
+      ],
+      "license": "MIT",
+      "license_family": "MIT",
+      "md5": "ba445bf767ae6f0d959ff2b40c20912b",
+      "name": "rich",
+      "noarch": "python",
+      "sha256": "2b26d58aa59e46f933c3126367348651b0dab6e0bf88014e857415bb184a4667",
+      "size": 184347,
+      "subdir": "noarch",
+      "timestamp": 1709150578093,
+      "version": "13.7.1"
+    }
+  },
+  // list of strings of keys (filenames) that were removed from either packages or packages.conda
+  "removed": [
+    "rich-10.15.1-pyhd8ed1ab_1.tar.bz2"
+  ]
+}
+```
+
+The `sha256` and `md5` from the original repodata fields are converted from their hex representation to bytes. 
+This is done to reduce the overall file size of the shards.
+
+Implementers SHOULD ignore unknown keys, this allows adding additional keys to the format in the future without breaking old versions of tools.
+
+Although these files can become relatively large (100s of kilobytes) typically for a large case (conda-forge) these files remaing very small, e.g. 100s of bytes to a couple of kilobytes.
+
+## <a id="authentication"></a>Authentication
```

Comment:
I'm -1 on adding this specific bit of authentication to the sharding proposal. We should be able to implement the sharding approach independently of the authentication mechanism. 

I don't see in some setups how we would be able to provide such an endpoint. This ties the CEP to a certain kind of authentication and removes the flexibility that we currently have in also using HTTP Basic Auth instead of token based auth.

Reply:
I am open to opening another CEP specifically for authentication of the sharded index but let me first explain the reasoning behind why we added this.

The reason we added this is that we have noticed that adding support for generic authentication methods provides a very significant overhead to fetching the individual shards and a large infrastructure cost. 

Take as an example our prefix.dev backend. When requesting private repodata a user requests `https://prefix.dev/mychannel/win-64/repodata.json`, our backend server authenticates the request using either a conda token, basic auth or an API key and redirects the request to a presigned-url. If we would do something similar for the shards each request would have to go through this dynamic end-point. This adds additional delay to the request and incurs an infrastructure cost due to the required compute. And we are talking about hundreds to thousands of requests when a user fetching repodata. We have noticed that this significantly impacts the performance.

Ideally, we want to have a solution where the "presigning" happens just once and we can reuse this information on subsequent requests.

I am open to alternative solutions, but I think being able to authenticate once for all subsequent requests is very important to how effective this proposal is. Standardizing this method is also important to ensure that every client will be able to access different server implementations. 

Then with regards to our specific solution using the `token` endpoint. The `token` endpoint only really needs to be dynamic when the authentication mechanism used is dynamic. In any other case, the token end-point could just be an empty static file. Basic authentication or even conda-token-based authentication can still be applied to both the individual shards and the shard index.

As an alternative though, we could also look at returning a set of headers from the initial request to the shard index instead of requesting a separate file.

FYI @jezdez

Reply:
I think what we are not making clear in the CEP is that "traditional" auth methods should continue to work just as well. For example, if the server uses basic HTTP authentication or conda token authentication, this should continue to work (and depending on the server implementation should also not negatively affect performance).

We are just interested in adding a mechanism to the protocol in order to facilitate really fast performance for large public channels.

I think we can make this an "add-on" behavior when the request to `repodata_shards.msgpack.zst` returns specific header values such as:

```
X-Shard-BaseUrl: https://fast.prefix.dev/conda-forge/
X-Shard-Token: pfx_secret_token
X-Shard-Token-ExpiresAt: TIMESTAMP
```

The client would then use these values for subsequent requests. If these headers are not returned, it will just use the "regular" URL with regular authentication.